pkg/controller/cainjector/controller.go

```diff
@@ -0,0 +1,211 @@
+	if splitPoint == -1 {
+		return types.NamespacedName{Name: nameStr}
+	}
+	return types.NamespacedName{Namespace: nameStr[:splitPoint], Name: nameStr[splitPoint+1:]}
+}
+
+// Reconcile attempts to ensure that a particular object has all the CAs injected that
+// it has requested.
+func (r *genericInjectReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.log.WithValues(r.resourceName, req.NamespacedName)
+
+	// fetch the target object
+	target := r.injector.NewTarget()
+	if err := r.Client.Get(ctx, req.NamespacedName, target.AsObject()); err != nil {
+		log.Error(err, "unable to fetch target object to inject into")
+		return ctrl.Result{}, err
+	}
+
+	// ensure that it wants injection
+	metaObj, err := meta.Accessor(target.AsObject())
+	if err != nil {
+		log.Error(err, "unable to get metadata for object")
+		return ctrl.Result{}, err
+	}
+	certNameRaw := metaObj.GetAnnotations()[WantInjectAnnotation]
+	if certNameRaw == "" {
+		log.V(1).Info("object does not want CA injection, skipping")
+		return ctrl.Result{}, nil
+	}
+
+	certName := splitNamespacedName(certNameRaw)
+	log = log.WithValues("certificate", certName)
+	if certName.Namespace == "" {
+		log.Error(nil, "invalid certificate name")
+		// don't return an error, requeuing won't help till this is changed
+		return ctrl.Result{}, nil
+	}
+
+	var cert certmanager.Certificate
+	if err := r.Client.Get(ctx, certName, &cert); err != nil {
+		log.Error(err, "unable to fetch associated certificate")
+		return ctrl.Result{}, err
+	}
+
+	// find all the points at which it can be injected
+	injectTargets := target.CAReferences()
+	if len(injectTargets) == 0 {
+		// if no injection targets, return early to avoid an extra update
+		return ctrl.Result{}, nil
+	}
+
+	// grab the associated secret, and ensure it's owned by the cert
+	secretName := types.NamespacedName{Namespace: cert.Namespace, Name: cert.Spec.SecretName}
+	log = log.WithValues("secret", secretName)
+	var secret corev1.Secret
+	if err := r.Client.Get(ctx, secretName, &secret); err != nil {
+		log.Error(err, "unable to fetch associated secret")
+		// don't requeue if we're just not found, we'll get called when the secret gets created
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	owner := OwningCertForSecret(&secret)
+	if owner == nil || *owner != certName {
+		log.Info("refusing to target secret not owned by certificate", "owner", metav1.GetControllerOf(&secret))
+		return ctrl.Result{}, nil
+	}
+
+	// inject the CA data
+	caData, hasCAData := secret.Data[certctrl.TLSCAKey]
+	if !hasCAData {
+		log.Error(nil, "certificate has no CA data")
+		// don't requeue, we'll get called when the secret gets updated
+		return ctrl.Result{}, nil
+	}
+
+	// actually do the injection
+	var lastErr error // continue on errors, but requeue if we had any
+	for _, injectable := range injectTargets {
+		injectable.SetCA(caData)
```

Comment:
Could you make the inject interface a bit cleaner by having SetCA set all the fields? It doesn't look like there's ever a case where you have one object with different CAs to their different targets.

Reply:
I originally had that functionality actually, and decided to scale down a bit since it didn't work well with non-service targets, and this is leftover from then. Good call, I'll fix the interface.